sdk/typescript/src/bcs/index.js

```diff
@@ -0,0 +1,663 @@
+            return new BN(value, 10);
+        }
+        /**
+         * Read U16 value from the buffer and shift cursor by 2.
+         * @returns {BN} 
+         */
+        read16() {
+            let value = this.dataView.getUint16(this.bytePosition, true);
+            this.shift(2);
+            return new BN(value, 10);
+        }
+        /**
+         * Read U32 value from the buffer and shift cursor by 4.
+         * @returns {BN} 
+         */
+        read32() {
+            let value = this.dataView.getUint32(this.bytePosition, true);
+            this.shift(4);
+            return new BN(value, 10);
+        }
+        /**
+         * Read U64 value from the buffer and shift cursor by 8.
+         * @returns {BN} 
+         */
+        read64() {
+            let value1 = this.read32(this.bytePosition);
+            let value2 = this.read32(this.bytePosition);
+            let result = value2.toString(16) + value1.toString(16).padStart(8, '0');
+            
+            return new BN(result, 16);
+        }
+        /**
+         * Read U128 value from the buffer and shift cursor by 16.
+         * @returns {BN} 
+         */
+        read128() {
+            let value1 = this.read64();
+            let value2 = this.read64();
+            let result = value2.toString(16) + value1.toString(16).padStart(8, '0');
+
+            return new BN(result, 16);
+        }
+        /**
+         * Read `num` number of bytes from the buffer and shift cursor by `num`.
+         * @param {Number} num Number of bytes to read.
+         * @returns {Uint8Array} Selected Buffer.
+         */
+        readXBytes(num) {
+            let start = this.bytePosition + this.dataView.byteOffset;
+            let value = new Uint8Array(this.dataView.buffer, start, num);
+
+            this.shift(num);
+
+            return value;
+        }
+        /**
+         * Read ULEB value - an integer of varying size. Used for enum indexes and
+         * vector lengths.
+         * @returns {Number} The ULEB value.
+         */
+        readULEB() {
+            let start = this.bytePosition + this.dataView.byteOffset;
+            let buffer = new Uint8Array(this.dataView.buffer, start);
+            let { value, length } = ulebDecode(buffer);
+
+            this.shift(length);
+
+            return value;
+        }
+        /**
+         * Read a BCS vector: read a length and then apply function `cb` X times
+         * where X is the length of the vector, defined in BCS bytes.
+         * @param {ReadVecCb} cb Callback to process elements of vector.
+         * @returns {Array<Any>} Array of the resulting values, returned by callback.
+         */
+        readVec(cb) {
+            let length = this.readULEB();  
+            let result = [];
+            for (let i = 0; i < length; i++) {
+                result.push(cb(this, i, length));
+            }
+            return result;
+        }
+    }
+    
+    /**
+     * The ReadVector callback.
```

Comment:
this documentation doesn't seem associated to any data

Reply:
It does in JSDoc. For proper generation of `.d.ts` I use JSDoc standard and define the Callback type with description of type parameters. But already rewriting everything to TS, so it won't matter.